### PR TITLE
Backports 0.15 pr24

### PR DESCRIPTION
**Pull request by PastaPastaPasta**

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -3618,50 +3655,20 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
                         strFailReason = _("Insufficient funds.");
                     }
 
-                    return false;
+                    return false;}
                 }
```
Comment (PastaPastaPasta) [inline src/wallet/wallet.cpp]: weird brace, possibly weird indentation

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -3670,7 +3677,7 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
                         {
                             nChangePosInOut = -1;
                             nFeeRet += nChange;
-                            reservekey.ReturnKey();
+
```
Comment (PastaPastaPasta) [inline src/wallet/wallet.cpp]: remove this line

File: src/wallet/rpcwallet.cpp [hunk 3]
```diff
@@ -1681,14 +1692,18 @@ UniValue listsinceblock(const JSONRPCRequest& request)
         return NullUniValue;
     }
 
-    if (request.fHelp)
```
Comment (PastaPastaPasta) [inline src/wallet/rpcwallet.cpp]: must have missed a PR...

Comment (UdjinM6) [reply src/wallet/rpcwallet.cpp]: yep, 10783

File: test/functional/multiwallet.py [hunk 4]
```diff
@@ -15,23 +20,49 @@ def __init__(self):
         self.extra_args = [['-wallet=w1', '-wallet=w2', '-wallet=w3']]
 
     def run_test(self):
+        self.stop_node(0)
+
+        # should not initialize if there are duplicate wallets
+        self.assert_start_raises_init_error(0, self.options.tmpdir, ['-wallet=w1', '-wallet=w1'], 'Error loading wallet w1. Duplicate -wallet filename specified.')
```
Comment (UdjinM6) [inline test/functional/multiwallet.py]: missing 10556

File: src/utilmoneystr.h [hunk 5]
```diff
@@ -14,6 +14,9 @@
 
 #include "amount.h"
 
+/* Do not use these functions to represent or parse monetary amounts to or from
+ * JSON but use AmountFromValue and ValueFromAmount for that.
```
Comment (UdjinM6) [inline src/utilmoneystr.h]: Should also fix few other things to follow this note, pls see https://github.com/UdjinM6/dash/commit/d326c8c94a8bb420e49d2dfc3c1ffb7aa39600c7.

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -3776,16 +3782,30 @@ bool CWallet::CreateTransaction(const std::vector<CRecipient>& vecSend, CWalletT
...
                     // the payees and not the change output.
-                    // TODO: The case where there is no change output remains
-                    // to be addressed so we avoid creating too small an output.
+
+                    // If we have no change and a big enough excess fee, then
+                    // try to construct transaction again only without picking
+                    // new inputs. We now know we only need the smaller fee
+                    // (because of reduced tx size) and so we should add a
+                    // change output. Only try this once.
+                    CAmount fee_needed_for_change = GetMinimumFee(change_prototype_size, currentConfirmationTarget, ::mempool, ::feeEstimator, nullptr, false /* ignoreGlobalPayTxFee */);
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/wallet/wallet.cpp]: 10598 is right here https://github.com/dashpay/dash/pull/3049/commits/c8e472c4b2d51108d430a66a9c7bf322bfc7b86f

[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/wallet/wallet.cpp]: oops, I mean 10589 "More economical fee estimates for RBF and RPC options to control" 🙈

**Comment by UdjinM6** [comment]
@PastaPastaPasta We could probably postpone missing PRs and just apply small patches to fix the build https://github.com/UdjinM6/dash/commits/pr3049 if backporting them here would cause some troubles for future backports you (presumably) prepared. (8f8399e1b42c0884f850b3fde50b36a9bbf8f023 would have to be reverted later though)

File: src/wallet/rpcwallet.cpp (hunk 3 above)
Comment (PastaPastaPasta) [reply src/wallet/rpcwallet.cpp]: added

File: test/functional/multiwallet.py (hunk 4 above)
Comment (PastaPastaPasta) [reply test/functional/multiwallet.py]: added, but LOTS of conflicts so likely some errors

**Comment by UdjinM6** [comment]
Well, that broke quite a few things 🙈Had to backport a part of 9672 https://github.com/bitcoin/bitcoin/pull/9672/commits/891c5eeec2dbf295d14e8f888de02637367dd930 -> https://github.com/UdjinM6/dash/commit/10d3639cf1c1b99820f4970282a76e6c441e65ed and then apply a bunch of fixes on top, pls see https://github.com/UdjinM6/dash/commits/pr3049. Not sure if I found all the missing parts though, will have a look again tomorrow if Travis isn't happy https://travis-ci.org/UdjinM6/dash/builds/569604147 (but at least tests are passing locally now).

**Comment by PastaPastaPasta** [comment]
see added, tests were still failing on your branch however.

File: src/wallet/rpcwallet.cpp [hunk 7]
```diff
@@ -1698,7 +1698,7 @@ UniValue listsinceblock(const JSONRPCRequest& request)
         return NullUniValue;
     }
 
-    if (request.fHelp || request.params.size() > 4)
+    if (request.fHelp || request.params.size() > 3)
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/wallet/rpcwallet.cpp]: squashed and pushed

File: src/wallet/rpcwallet.cpp [hunk 8]
```diff
@@ -799,11 +799,11 @@ UniValue getbalance(const JSONRPCRequest& request)
     const std::string* account = account_param != "*" ? &account_param : nullptr;
 
     int nMinDepth = 1;
-    if (request.params.size() > 1)
+    if (!request.params[1].isNull())
         nMinDepth = request.params[1].get_int();
     bool fAddLocked = (request.params.size() > 2 && request.params[2].get_bool());
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/wallet/rpcwallet.cpp]: I feel as though we should be able to do a scripted diff somehow...

File: src/wallet/rpcwallet.cpp [hunk 9]
```diff
@@ -964,7 +986,7 @@ UniValue sendmany(const JSONRPCRequest& request)
 
     if (request.fHelp || request.params.size() < 2 || request.params.size() > 8)
```
[the remark this answers is not in the document]
Comment (PastaPastaPasta) [reply src/wallet/rpcwallet.cpp]: see 806da3c6e

**Comment by UdjinM6** [comment]
Pls also see https://github.com/UdjinM6/dash/commits/pr3049 for more fixes, specifically 4ccac04d2bf8d7ecf95e663d138ca10c28fc12a4, dc83c35f029b61c45ec867396efcf77b88662965, 55fe1c59514f7d3ab61ba64005d6db839af3a6d7).

**Comment by PastaPastaPasta** [comment]
all applied

File: src/wallet/rpcwallet.cpp (hunk 8 above)
[the remark this answers is not in the document]
Comment (UdjinM6) [reply src/wallet/rpcwallet.cpp]: `!isNull()` is needed when you need to access it via named params. Should probably be applied to Dash specific ones but it's probably ok to do this in a separate PR.

**Pull request merged**